Added delay utility

sseifert-akamai opened the pull request:
Hi @ynohat , I added a delay utility to utils.libsonnet and an example test case to akamai.jsonnet. The delay utility allows delay of test execution and can be defined in ms.

sseifert-akamai committed:
Added delay utility to utils.libsonnet and example test case to akamai.jsonnet

ynohat reviewed (changes requested):
@sve-sei the change looks great, and thanks for contributing it!

I understand the example makes a lot of sense for testing cache expiry in Akamai. However it makes it a little bit less discoverable, since the utility function `delay` actually has nothing to do with Akamai at all.

Can we consider adding a `example/delaying-requests.jsonnet` suite instead?
The pre script could store the current time, introduce the delay, and the test script could check that the current time is larger than the delay.

I also think I would prefer the function to live outside utils, which is really a "jsonnet utils" thing more than a "javascript generator" thing. I would suggest moving the delay function to `src/delay.libsonnet` and changing the API to something like this:

```
test.case {
  ...
} + test.delay(3000)
```

Finally, is there a reason for using the full form of test.GET in the example? I'd encourage the use of just the URL if it's not necessary, to keep things compact - here one might think it is required to use the full form.

sseifert-akamai committed:
Modified delay function code and test case

sseifert-akamai committed:
Merge branch 'delay-function' into delay-utility

sseifert-akamai committed:
Updated code for delay function and test case

sseifert-akamai commented:
@ynohat Thank you for your comments and suggestions. 

I have updated the code and split the example in example/delaying-requests.jsonnet as you suggested and simplified the test.GET. I also updated the code in case.libsonnet so that it is possible to add an optional "delay" parameter to the test case with the required timeout that will be executed before the request is sent.

ynohat commented on example/akamai.jsonnet:
```diff
@@ -47,6 +47,24 @@ test.suite {
+    test.case {
+      name: 'Delay of test execution by 3 seconds',
+      pre:: [test.utils.delay(3000)], // Add a delay of 3s before the test is executed
+      request: test.GET({
```
sorry to bother you with this, but can you please replace this syntax with:

`      request: test.POST('https://{{httpbin}}/response-headers?x-cache=TCP_MISS')`

The full request syntax is useful but not necessary here, and I would prefer folks copy pasted simpler examples.

sseifert-akamai replied:
I just created a new commit. Please check again.

ynohat commented on src/utils.libsonnet:
```diff
@@ -9,5 +9,7 @@
       local fmtParam(p) = '%s=%s' % [self.escapeString(p), self.escapeString(params[p])];
       std.join('&', std.map(fmtParam, std.objectFields(params))),
+
+    delay(ms):: 'setTimeout(function(){}, %s);' % std.manifestJson(ms)
```
please remove this from utils, since we agreed that it was more of a script than a jsonnet utility

sseifert-akamai replied:
I just created a new commit. Please check again.

ynohat commented on example/akamai.jsonnet:
```diff
@@ -47,6 +47,24 @@ test.suite {
+    // This can be useful, if previous tests or actions require a delay before the next the can be executed
+    test.case {
+      name: 'Delay of test execution by 3 seconds',
+      pre:: [test.utils.delay(3000)], // Add a delay of 3s before the test is executed
```
please simply use the flag

sseifert-akamai replied:
I just created a new commit. Please check again.

ynohat commented on example/delaying-requests.jsonnet:
```diff
@@ -0,0 +1,23 @@
+
+test.suite {
+  name: 'Delaying tests',
+  item:
```
I think the syntax here will fail at compile time

ynohat replied:
... it will not, simply did not know myself that postman accepted something else than a list in the spec for that field !

ynohat reviewed (changes requested):
we're nearly there!

sseifert-akamai committed:
Reverted changes in example/akamai.jsonnet and src/utils.libsonnet

sseifert-akamai committed:
Added delaying-requests.jsonnet to main.jsonnet test suite
